app/eth-watcher.hoon

```diff
@@ -0,0 +1,556 @@
+    |=  [b=bone s=ship p=path]
+    ^-  (unit bone)
+    ?>  ?=([@ *] p)
+    ?:(=(name i.p) `b ~)
+  ::
+  ++  fan-diff
+    |=  dif=update
+    %-  put-moves
+    %+  turn  get-subscribers
+    |=  b=bone
+    ^-  move
+    [b %diff %eth-watcher-update dif]
+  ::
+  ++  cancel-subscribers
+    %-  put-moves
+    %+  turn  get-subscribers
+    |=(b=bone [b %quit ~])
+  ::
+  ::  +|  configuration
+  ::
+  ::  +init: set up eye
+  ::
+  ++  init
+    |=  [nom=^name =config]
+    ^+  +>
+    =.  name  nom
+    =.  eye
+      %*(. *^eye - config, last-heard-block from-block.config)
+    get-latest-block
+  ::
+  ::  +|  catch-up-operations
+  ::
+  ::  +get-latest-block
+  ::
+  ::    Get latest block from eth node and compare to our own latest block.
+  ::    Get intervening blocks in chunks until we're caught up, then set
+  ::    up a filter going forward.
+  ::
+  ++  get-latest-block
+    =>  cancel-wait-poll
+    (put-rpc-request /catch-up/block-number `'block number' %eth-block-number ~)
+  ::
+  ::  +catch-up: get next chunk
+  ::
+  ++  catch-up
+    |=  from-block=@ud
+    ^+  +>
+    ?:  (gte from-block latest-block.eye)
+      new-filter
+    =/  next-block  (min latest-block.eye (add from-block 5.760))  ::  ~d1
+    ~?  debug=|
+      [%catching-up from=from-block to=latest-block.eye]
+    %-  put-rpc-request
+    :+  /catch-up/step/(scot %ud from-block)/(scot %ud next-block)
+      `'catch up'
+    :*  %eth-get-logs
+        `number+from-block
+        `number+next-block
+        contracts.eye
+        topics.eye
+    ==
+  ::
+  ::  +|  filter-operations
+  ::
+  ::  +new-filter: request a new polling filter
+  ::
+  ::    Listens from the last-heard block onward.
+  ::
+  ++  new-filter
+    %-  put-rpc-request
+    :+  /filter/new  `'new filter'
+    ^-  request:rpc
+    :*  %eth-new-filter
+        `number+last-heard-block.eye
+        ?~(to-block.eye ~ `number+u.to-block.eye)
+        contracts.eye
+        topics.eye
+    ==
+  ::
+  ::  +read-filter: get all events the filter captures
+  ::
+  ++  read-filter
+    %-  put-rpc-request
+    :+  /filter/logs  `'filter logs'
+    [%eth-get-filter-logs filter-id.eye]
+  ::
+  ::  +poll-filter: get all new events since last poll (or filter creation)
+  ::
+  ++  poll-filter
+    ?:  =(0 filter-id.eye)
+      ~&  %no-filter-bad-poll
+      .
+    %-  put-rpc-request
+    :+  /filter/changes  `'poll filter'
+    [%eth-get-filter-changes filter-id.eye]
+  ::
+  ::  +wait-poll: remind us to poll in four minutes
+  ::
+  ::    Four minutes because Ethereum RPC filters time out after five.
+  ::    We don't check for an existing timer or clear an old one here,
+  ::    sane flows shouldn't see this being called superfluously.
+  ::
+  ++  wait-poll
+    =+  wen=(add now ~m4)
+    %-  put-move(poll-timer.eye wen)
+    [%wait name^/poll wen]
+  ::
+  ::  +cancel-wait-poll: remove poll reminder
+  ::
+  ++  cancel-wait-poll
+    ?:  =(*@da poll-timer.eye)  ..cancel-wait-poll
```

Comment:
If something is nullable, it should be a `unit`.  Otherwise it's easy to assume it exists when it doesn't.